core-backend/sandbox/src/funcs.rs

```diff
@@ -344,15 +340,13 @@ where
         at: u32,
         len: u32,
     ) -> Result<&'_ [u8], FuncError<<E as Ext>::Error>> {
-        let msg = ext.read().map_err(FuncError::Core)?;
+        let msg = ext.read(at, len).map_err(FuncError::Core)?;
 
-        let last_idx = at
-            .checked_add(len)
-            .ok_or_else(|| FuncError::ReadLenOverflow(at, len))?;
+        // 'at' and 'len' correct and saturation checked in Ext::read
+        debug_assert!(at.checked_add(len).is_some());
+        let last_idx = at + len;
 
-        if last_idx as usize > msg.len() {
-            return Err(FuncError::ReadWrongRange(at..last_idx, msg.len() as u32));
-        }
+        debug_assert!(last_idx as usize >= msg.len());
```

Comment:
Isn't it?
```suggestion
        debug_assert!(last_idx as usize <= msg.len());
```

Reply:
Also make sure in validation `at <= last_idx`

Reply:
done